### ISSUE TITLE
无法复现GSM8K上的结果

### ISSUE THREAD

**Issue by greycooker** [mrflogs/LoRA-Pro#7]
作者您好！我这边遇到了与之前的issue一样的问题https://github.com/mrflogs/LoRA-Pro/issues/3
Evaluate seed 0 results:
Test samples 1319
Test samples 1319
Final Accuracy: 0.1516300227445034
想问一下代码的bug已经修复了吗？我使用的模型是modelscope上的Llama-2-7b-hf https://www.modelscope.cn/models/shakechen/Llama-2-7b-hf/files

**Comment by mrflogs** [comment]
我们使用的是huggingface上比较早期版本的llama。服务器目前没空，后续有空机器了我会再更新。

transformers.LlamaForCausalLM.from_pretrained("meta-llama/Llama-2-7b-hf", revision="637a748546bb9abca62b0684183cc362bc1ece6d")

**Comment by LZY-the-boys** [comment]
Hi, you can try this PR offered by me https://github.com/mrflogs/LoRA-Pro/pull/8, 
it is because batch generation they don't input the attention mask

**Comment by greycooker** [comment]
Thanks!!!

**Comment by greycooker** [comment]
这边经过修改模型版本以后rslora-pro正确率可以达到54，与论文一致。
但是我发现lora和rslora的正确率也有52，而不是论文里的42，而且看训练loss其实rslora和rslora-pro相差不大
**lora代码修改**：
`    def create_optimizer_and_scheduler(self, num_training_steps: int):
        if "pro" in self.args1.lora:
            self.optimizer = peta.optim.AdamW(self.model.named_parameters(), lr=self.args.learning_rate, scaling_factor=self.scaling_factor, betas=(0.9, 0.999), weight_decay=self.args.weight_decay, mode="efficient", X_mode='sylvester')
        else:
            # lora-pro
            self.optimizer = AdamW(
                self.model.parameters(),  
                lr=self.args.learning_rate,
                betas=(0.9, 0.999),
                weight_decay=self.args.weight_decay,
            )
        self.create_scheduler(num_training_steps=num_training_steps, optimizer=self.optimizer)`

**测试脚本**：
# rslora-pro
echo "rslora-pro"
torchrun --nproc_per_node=8 minimal_lora_llama2_math_transformers.py --lora rslora-pro --seed 0
torchrun --nproc_per_node=8 evaluation/eval_llama-2_math_multi_gpus.py --lora rslora-pro --seed 0
# lora
echo "lora"
torchrun --nproc_per_node=8 minimal_lora_llama2_math_transformers.py --lora lora --seed 0
torchrun --nproc_per_node=8 evaluation/eval_llama-2_math_multi_gpus.py --lora lora --seed 0
# rslora
echo "rslora"
torchrun --nproc_per_node=8 minimal_lora_llama2_math_transformers.py --lora rslora --seed 0
torchrun --nproc_per_node=8 evaluation/eval_llama-2_math_multi_gpus.py --lora rslora --seed 0
# lora-pro
echo "lora-pro"
torchrun --nproc_per_node=8 minimal_lora_llama2_math_transformers.py --lora lora-pro --seed 0
torchrun --nproc_per_node=8 evaluation/eval_llama-2_math_multi_gpus.py --lora lora-pro --seed 0

**训练loss**
<img width="1267" alt="b8065092f05512a5e0577063953aca7b" src="https://github.com/user-attachments/assets/c24bbcaa-cf0c-450f-bfa9-4e88b8e30f0d" />

**Comment by LZY-the-boys** [comment]
Hi，you should turn zero2 off to truly use lora-pro，please view zero2 issue in #6 .

**Comment by greycooker** [comment]
> Hi，you should turn zero2 off to truly use lora-pro，please view zero2 issue in #6 .

Thank you for your reminder, I will try to turn off zero2. I am very curious if the experiment in the paper is based on this repo and turn zero2 on, will the experimental results in the paper still be credible?

**Comment by greycooker** [comment]
<img width="375" alt="5151e7b8123cedc9ca00d325005ef13b" src="https://github.com/user-attachments/assets/8120ac94-e6b0-4d66-9510-c500486fc60a" />
https://openreview.net/forum?id=gTwRMU3lJ5在openreview上LoRA-Pro看起来效果非常棒，想请教一下这里的实验是怎么做的？

**Comment by mrflogs** [comment]
We have resolved this issue, and the results have been re-run and validated accordingly.

In LoRA-Pro, to ensure compatibility with DeepSpeed, we've integrated the Adam optimization process directly into DeepSpeed (in [DeepSpeed-0.15.1/deepspeed/runtime/zero/stage_1_and_2.py](https://github.com/mrflogs/LoRA-Pro/blob/main/DeepSpeed-0.15.1/deepspeed/runtime/zero/stage_1_and_2.py?rgh-link-date=2025-11-17T06%3A51%3A27.000Z#L1828)). Therefore, in the TrainingArguments, you need to set the optimizer to "sgd" to prevent parameters from being updated twice.
